deob-bytecode/src/main/kotlin/org/openrs2/deob/bytecode/transform/StringDecryptionTransformer.kt

```diff
@@ -0,0 +1,183 @@
+
+                // step 1: identify the decryption methods
+                val inner = clinit.instructions.iterator().asSequence().filterIsInstance<MethodInsnNode>()
+                    .filter {
+                        it.owner == clazz.name && it.desc == "(Ljava/lang/String;)[C" &&
+                            it.next?.opcode == Opcodes.INVOKESTATIC
+                    }.firstOrNull()?.let {
+                        clazz.methods.find { method -> method.name == it.name && method.desc == it.desc }
+                    }
+                val outer = clinit.instructions.iterator().asSequence().filterIsInstance<MethodInsnNode>()
+                    .filter {
+                        it.owner == clazz.name && it.desc == "([C)Ljava/lang/String;" &&
+                            it.previous?.opcode == Opcodes.INVOKESTATIC
+                    }.firstOrNull()?.let {
+                        clazz.methods.find { method -> method.name == it.name && method.desc == it.desc }
+                    }
+
+                if (inner == null || outer == null) continue
+
+                // step 2: get the xor table from the outer method
+                val switch: TableSwitchInsnNode = outer.instructions.iterator().asSequence()
+                    .filterIsInstance<TableSwitchInsnNode>().firstOrNull() ?: continue
+
+                val xorLookup = ArrayList<Number>()
+                for (label in switch.labels) {
+                    xorLookup.add(label.next.intConstant ?: continue)
+                }
+                xorLookup.add(switch.dflt.next.intConstant ?: continue)
+
+                // step 3: decrypt all strings in the initializer
+                val strings = ArrayList<String>()
+                var field: FieldNode? = null
+                var fieldInit: AbstractInsnNode? = null
+                for (insn in clinit.instructions) {
+                    if (insn.opcode != Opcodes.INVOKESTATIC) continue
+                    val invoke = insn as MethodInsnNode
+                    if (invoke.owner != clazz.name || invoke.name != outer.name || invoke.desc != outer.desc) continue
+
+                    val ldc = invoke.previous.previous as LdcInsnNode
+                    val str = ldc.cst as String
+                    val decrypted = decryptString(str, xorLookup)
+
+                    ldc.cst = decrypted
+                    strings.add(decrypted)
+                    stringsDecrypted++
+
+                    if (field == null) {
+                        if (invoke.next?.opcode == Opcodes.AASTORE && invoke.next?.next?.opcode == Opcodes.PUTSTATIC) {
+                            // this one is the static String[] field
+                            val put = invoke.next.next as FieldInsnNode
+                            if (put.owner == clazz.name && put.desc == "[Ljava/lang/String;") {
+                                field = clazz.fields.find { it.name == put.name }
```

Comment:
I think this should check the descriptor too, as (iirc?) fields can overload each other at the bytecode level.